privacyidea/models.py

```diff
@@ -1050,9 +1052,69 @@ def __init__(self, resolver_id=None, realm_id=None,
                                  .first().id
 
 
+class TokenOwner(MethodsMixin, db.Model):
+    """
+    This tables stores the owner of a token.
+    A token can be assigned to several users.
+    """
+    __tablename__ = 'tokenowner'
+    id = db.Column(db.Integer(), Sequence("tokenowner_seq"), primary_key=True,
+                   nullable=True)
+    token_id = db.Column(db.Integer(), db.ForeignKey('token.id'))
+    token = db.relationship('Token', lazy='joined', backref='token_list')
+    resolver = db.Column(db.Unicode(120), default=u'', index=True)
+    resolver_type = db.Column(db.Unicode(120), default=u'')
+    user_id = db.Column(db.Unicode(320), default=u'', index=True)
+    realm_id = db.Column(db.Integer(), db.ForeignKey('realm.id'))
+    realm = db.relationship('Realm', lazy='joined', backref='realm_list')
```

Comment:
I renamted this to "tokenowner". Not quite sure, in which commit.

Well, the only question remains, what we really should do with the *laziness*

Reply:
I think Lazy loading (lazy='select' https://docs.sqlalchemy.org/en/latest/orm/loading_relationships.html) is good if
1. you usually do not need the relation or need it later or
2. if you related object is very big.

The lazy=select issues a 2nd select statement for the relation. It many cases it is probably better to jave *one* joines select statement than 2 select statements.

So we need to look at, where we need the relationship immediately or where we need the relationship later - we have no *big* object (So this reason for lazy loading does not count).

# class Token

*info* and *owner*. I think we *mostly* need the additional tokeninfo and we *always* need the owner of the token, if we are handling with tokens.
So my guess would be to change these two
https://github.com/privacyidea/privacyidea/blob/1288/more-users-per-token/privacyidea/models.py#L180
https://github.com/privacyidea/privacyidea/blob/1288/more-users-per-token/privacyidea/models.py#L184
actually to 'joined'.

# TokenInfo

oups, we have a double realtion here, 
https://github.com/privacyidea/privacyidea/blob/1288/more-users-per-token/privacyidea/models.py#L642
although we only have an 1:m relation (one token, many attributes)
But also in this case, I would prefer a 'joined' laziness.
I will check, if we can drop the info in the Token.

# Resolver / ResolverConfig

https://github.com/privacyidea/privacyidea/blob/1288/more-users-per-token/privacyidea/models.py#L921
https://github.com/privacyidea/privacyidea/blob/1288/more-users-per-token/privacyidea/models.py#L960token/privacyidea/models.py#L921

Here we also have a probably unnessary double relation.
But this needs to be joined, since a resolver is never used without its config.

# ResolverRealm

is used for the n:m relation to bind any possible resolver to any possible realm.
https://github.com/privacyidea/privacyidea/blob/1288/more-users-per-token/privacyidea/models.py#L1019
https://github.com/privacyidea/privacyidea/blob/1288/more-users-per-token/privacyidea/models.py#L1024

To my understanding it is always neccessary to do a joined laziness.

# TokenOwner

I am not sure, if we always need the realm, when we want to find, which user belongs to a token.
https://github.com/privacyidea/privacyidea/blob/1288/more-users-per-token/privacyidea/models.py#L1068

Usually we get the user and want to find his tokens, so we should take a look at the ``get_tokens(user)``.
But from a data size perspective, this join would not hurt, since the realm is only a name.

# TokenRealm

https://github.com/privacyidea/privacyidea/blob/1288/more-users-per-token/privacyidea/models.py#L1126
https://github.com/privacyidea/privacyidea/blob/1288/more-users-per-token/privacyidea/models.py#L1130

This connects any token to any realm.
I am not sure, if this needs to be joined. At a first glance it looks like this information is not necessariely needed durch authentication. The realm would already be available via the tokenowner.

# Eventhandler

https://github.com/privacyidea/privacyidea/blob/1288/more-users-per-token/privacyidea/models.py#L1656
https://github.com/privacyidea/privacyidea/blob/1288/more-users-per-token/privacyidea/models.py#L1660
https://github.com/privacyidea/privacyidea/blob/1288/more-users-per-token/privacyidea/models.py#L1770
https://github.com/privacyidea/privacyidea/blob/1288/more-users-per-token/privacyidea/models.py#L1818

Eventhandler depend on the configuration and conditions.
So we should fetch all information from the start in a joined laziness.

# Periodic Tasks

https://github.com/privacyidea/privacyidea/blob/1288/more-users-per-token/privacyidea/models.py#L2615
https://github.com/privacyidea/privacyidea/blob/1288/more-users-per-token/privacyidea/models.py#L2618

You say ;-)

Reply:
We should define a nice test scenario for investigating this.
But we might also put the question of laziness into another issue and PR.

Reply:
THere are a lot of comment/conversations. Could you please resolver the conversation? Thanks.